README.md
# sbt-compile-analyzer #

Welcome to sbt-compile-analyzer!

This program allows the extraction of compiler information for the single phases of the `scalac` compilation process. Therefore, the program needs the information that is generated by the `scalac` compiler during its compilation of a software program.

The content of the data file consists of the `scalac` compiler information that can be generated in the following manner. 

Add the following `scalac` compiler option to the build.sbt of the software program that should be analyzed:
  
`scalacOptions ++= Seq("-verbose")`

Run the compilation of the project and generate the data file:
`sbt clean compile > data-file.txt`

Switch to this project and execute one of the following commands:

`
sbt
run absolute/path/to/data-file.txt
`

or

`sbt "runMain com.wegtam.sbt.compile.analyzer.Analyzer absolute/path/to/data-file.txt"`

## Example output

The following two tables represent the result of this program to itself.

<pre><code>

===========================================
Overview of the single phases
===========================================
| ID  | Phase                | Took (ms)  |
| 1   | ParserPhase          | 97         |
| 2   | NamerPhase           | 106        |
| 3   | PackageObjectsPhase  | 0          |
| 4   | TyperPhase           | 1794       |
| 5   | PatmatPhase          | 714        |
| 6   | SuperAccessorsPhase  | 58         |
| 7   | ExtMethodsPhase      | 11         |
| 8   | PicklerPhase         | 52         |
| 9   | RefChecksPhase       | 195        |
| 10  | UncurryPhase         | 137        |
| 11  | TailCallsPhase       | 43         |
| 12  | SpecializePhase      | 365        |
| 13  | ExplicitOuterPhase   | 116        |
| 14  | ErasurePhase         | 364        |
| 15  | PostErasurePhase     | 20         |
| 17  | LambdaLiftPhase      | 54         |
| 18  | ConstructorsPhase    | 43         |
| 19  | FlattenPhase         | 26         |
| 20  | MixinPhase           | 46         |
| 21  | CleanupPhase         | 17         |
| 22  | DelambdafyPhase      | 46         |
| 26  | Total                | 6024       |

====================================================
Overview of the single phases (ordered by used time)
====================================================
| ID  | Phase                | Took (ms)  | %      |
| 26  | Total                | 6024       | 100    |
| 4   | TyperPhase           | 1794       | 29,78  |
| 5   | PatmatPhase          | 714        | 11,85  |
| 12  | SpecializePhase      | 365        | 6,06   |
| 14  | ErasurePhase         | 364        | 6,04   |
| 9   | RefChecksPhase       | 195        | 3,24   |
| 10  | UncurryPhase         | 137        | 2,27   |
| 13  | ExplicitOuterPhase   | 116        | 1,93   |
| 2   | NamerPhase           | 106        | 1,76   |
| 1   | ParserPhase          | 97         | 1,61   |
| 6   | SuperAccessorsPhase  | 58         | 0,96   |
| 17  | LambdaLiftPhase      | 54         | 0,9    |
| 8   | PicklerPhase         | 52         | 0,86   |
| 20  | MixinPhase           | 46         | 0,76   |
| 22  | DelambdafyPhase      | 46         | 0,76   |
| 11  | TailCallsPhase       | 43         | 0,71   |
| 18  | ConstructorsPhase    | 43         | 0,71   |
| 19  | FlattenPhase         | 26         | 0,43   |
| 15  | PostErasurePhase     | 20         | 0,33   |
| 21  | CleanupPhase         | 17         | 0,28   |
| 7   | ExtMethodsPhase      | 11         | 0,18   |
| 3   | PackageObjectsPhase  | 0          | 0      |

</code></pre>

## Contribution policy ##

Contributions via GitHub pull requests are gladly accepted from their original author. Along with
any pull requests, please state that the contribution is your original work and that you license
the work to the project under the project's open source license. Whether or not you state this
explicitly, by submitting any copyrighted material via pull request, email, or other means you
agree to license the material under the project's open source license and warrant that you have the
legal authority to do so.

## License ##

This code is open source software licensed under the
[AGPL-3.0](https://www.gnu.org/licenses/agpl.html) license.


CHANGELOG.md
# ChangeLog for the SBT Compile Phase Analyzer

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## Conventions when editing this file.

Please follow the listed conventions when editing this file:

* one subsection per version
* reverse chronological order (latest entry on top)
* write all dates in iso notation (`YYYY-MM-DD`)
* each version should group changes according to their impact:
    * `Added` for new features.
    * `Changed` for changes in existing functionality.
    * `Deprecated` for once-stable features removed in upcoming releases.
    * `Removed` for deprecated features removed in this release.
    * `Fixed` for any bug fixes.
    * `Security` to invite users to upgrade in case of vulnerabilities.

## Unreleased

## 2017-06-17

- Added some more type annotations
- Updated the example in the README

## 2017-06-15

### Added

- initial Release

src/main/scala/com/wegtam/sbt/compile/analyzer/Helper.scala
/*
 * Copyright (C) 2017  André Schütz
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.wegtam.sbt.compile.analyzer

import java.text.DecimalFormat

import adt.{ ScalacPhase, ScalacPhaseName }
import adt.ScalacPhaseName.{ Loader, Total }
import adt.ScalacPhaseValue.ScalacPhaseValueLoader

/**
  * A few helper methods for the project.
  */
@SuppressWarnings(Array("org.wartremover.warts.Any"))
object Helper {

  /**
    * An ugly simple fast possibility to visualize the information of the
    * extracted phases from the data file.
    *
    * @param r Sequence of phase information
    */
  def print(r: Seq[ScalacPhase]): Unit = {
    println("\n\n===========================================")
    println("Overview of the single phases")
    println("===========================================")
    println("| %-3s | %-20s | %-10s |".format("ID", "Phase", "Took (ms)"))
    r.sortBy(_.id).drop(1).foreach { phase =>
      println("| %-3d | %-20s | %-10d |".format(phase.id, phase.name, phase.took))
    }

    println("\n\n====================================================")
    println("Overview of the single phases (ordered by used time)")
    println("====================================================")
    println("| %-3s | %-20s | %-10s | %-6s |".format("ID", "Phase", "Took (ms)", "%"))
    val sortedByTime: Seq[ScalacPhase] =
      r.sortBy(_.id).drop(1).sortBy(_.took)(Ordering[Long].reverse)
    val totalTime: Long =
      sortedByTime.find(_.id == ScalacPhaseName.phaseId(Total).getOrElse(-1)).fold(0L)(t => t.took)
    r.sortBy(_.id).drop(1).sortBy(_.took)(Ordering[Long].reverse).foreach { phase =>
      val tookPercentage: BigDecimal =
        if (phase.took > 0)
          BigDecimal(phase.took.toDouble / totalTime.toDouble * 100)
            .setScale(2, BigDecimal.RoundingMode.HALF_UP)
        else 0
      println(
        "| %-3d | %-20s | %-10d | %-6s |"
          .format(phase.id,
                  phase.name,
                  phase.took,
                  new DecimalFormat("#0.##").format(tookPercentage))
      )
    }

  }

  /**
    * Helper function to print the information from the loader.
    * Currently unused.
    *
    * @param r Sequence of phase information
    */
  private def printLoaderData(r: Seq[ScalacPhase]): Unit = {
    // Loader
    val loaderPhase: Option[ScalacPhase] = r.find(_.name == Loader)
    loaderPhase.fold(println(s"No data for ${Loader.toString}")) { loader =>
      println("============================================================")
      println(s"${Loader.toString} (ID: ${loader.id}) took ${loader.took}ms")
      println("============================================================")
      println("\nSlowest loading times:")
      loader.values
        .sortBy {
          case ScalacPhaseValueLoader(desc1: String, desc2: String, value: String, took: Long) =>
            took
          case _ => -1
        }(Ordering[Long].reverse)
        .takeWhile {
          case ScalacPhaseValueLoader(desc1: String, desc2: String, value: String, took: Long) =>
            took > 10
          case _ => true
        }
        .foreach {
          case ScalacPhaseValueLoader(desc1: String, desc2: String, value: String, took: Long) =>
            println(s"$desc1, $desc2, $value, ${took}ms")
          case _ => ""
        }
    }
  }

}


src/main/scala/com/wegtam/sbt/compile/analyzer/Analyzer.scala
/*
 * Copyright (C) 2017  André Schütz
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.wegtam.sbt.compile.analyzer

import com.wegtam.sbt.compile.analyzer.adt.ScalacPhaseName.Loader
import com.wegtam.sbt.compile.analyzer.adt.ScalacPhaseValue.ScalacPhaseValueLoader
import com.wegtam.sbt.compile.analyzer.adt.{ ScalacPhase, ScalacPhaseName }

import scala.io.Source
import scala.util.matching.Regex

/**
  * This program allows the extraction of compiler information for the
  * single phases of the `scalac` compilation process. Therefore, the
  * program needs the information that is generated by the `scalac` compiler
  * during its compilation of a software program.
  *
  * The content of the data file consists of the `scalac` compiler information
  * that can be generated in the following manner.
  *
  * Add the following `scalac` compiler option to the build.sbt:
  * `scalacOptions ++= Seq("-verbose")`
  *
  * Run the compilation of your project and generate the data file:
  * `sbt clean compile > data-file.txt`
  *
  * Switch to this project and execute one of the following commands:
  *
  * `
  *  sbt
  *  run absolute/path/to/data/file.txt
  * `
  * or
  *
  * `sbt "runMain Analyzer absolute/path/to/data/file.txt"`
  */
object Analyzer {

  /**
    * Main entry point for the program.
    *
    * @param args The arguments that can be provided during runtime.
    */
  def main(args: Array[String]): Unit = {
    // The provided filepath to the data file
    // Yes, there is no check for non-existence :-)
    val fp: String = args(0)

    val phases: Seq[ScalacPhase] = Analyzer(fp).analyze()
    Helper.print(phases)
  }
}

/**
  * This class provides functionality to extract the information from
  * the provided data file.
  *
  * @param filepath The path to the data file.
  */
final case class Analyzer(filepath: String) {

  /**
    * Start the analysis process and return the extracted information
    * for the single phases.
    *
    * @return Sequence of information for the single phases.
    */
  def analyze(): Seq[ScalacPhase] = {
    val lines: Seq[String] = readFile()
    phases(lines)
  }

  /**
    * Create the information for the single phases that can be
    * extracted from the content of the data file.
    *
    * @param lines  The raw lines of the data file.
    * @return Sequence of phase information that could be extracted.
    */
  def phases(lines: Seq[String]): Seq[ScalacPhase] = {
    // Loader
    val loaderValuesRegex: Regex = ".*\\[loaded (\\w+) (\\w+) (.*) in (\\d+)ms].*".r
    val lvr: Seq[ScalacPhaseValueLoader] = lines.flatMap {
      case loaderValuesRegex(desc1: String, desc2: String, value: String, took: String) =>
        Option(ScalacPhaseValueLoader(desc1, desc2, value, took.toLong))
      case _ => None
    }
    val loaderPhase: ScalacPhase = ScalacPhase(
      ScalacPhaseName.phaseId(Loader).getOrElse(-1),
      Loader,
      -1L,
      lvr
    )

    // load all phases
    val allPhasesRegex: Regex = ".*\\[(\\w+) in (\\d+)ms\\].*".r
    val phases: Seq[ScalacPhase] = lines.flatMap {
      case allPhasesRegex(phase: String, took: String) =>
        ScalacPhaseName
          .valueOf(phase)
          .fold[Option[ScalacPhase]](None)(
            p =>
              Option(
                ScalacPhase(ScalacPhaseName.phaseId(p).getOrElse(-1), p, took.toLong, Seq.empty)
            )
          )
      case _ => None
    }

    phases ++ Seq(loaderPhase)
  }

  /**
    * Read the file and return a sequence of the contained lines.
    * @return Sequence of the file content
    */
  def readFile(): Seq[String] = Source.fromFile(filepath).getLines().toSeq
}


src/main/scala/com/wegtam/sbt/compile/analyzer/adt/ScalacPhase.scala
/*
 * Copyright (C) 2017  André Schütz
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.wegtam.sbt.compile.analyzer.adt

/**
  * A phase contains all information that are available regarding the
  * Scalac compilation of this phase.
  *
  * @param id     The ID of the phase. Normally the order at which the phase occurs.
  * @param name   The name of the phase.
  * @param took   How long the phase took in Milliseconds.
  * @param values Additional information regarding this phase.
  */
case class ScalacPhase(
    id: Int,
    name: ScalacPhaseName,
    took: Long,
    values: Seq[ScalacPhaseValue]
)


src/main/scala/com/wegtam/sbt/compile/analyzer/adt/ScalacPhaseValue.scala
/*
 * Copyright (C) 2017  André Schütz
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.wegtam.sbt.compile.analyzer.adt

/**
  * Sealed trait to collect the different values that can be connected
  * to a phase and extracted from the data.
  */
sealed trait ScalacPhaseValue

object ScalacPhaseValue {

  /**
    * General case where just a String value is available.
    *
    * @param value  Value for the phase.
    */
  case class ScalacPhaseValueGeneral(value: String) extends ScalacPhaseValue

  /**
    * Values of the loader phase.
    *
    * @param desc1  Describes the `value`.
    * @param desc2  Describes the `value`.
    * @param value  Value for the phase.
    * @param took   How long this value took in Milliseconds.
    */
  case class ScalacPhaseValueLoader(desc1: String, desc2: String, value: String, took: Long)
      extends ScalacPhaseValue

}


src/main/scala/com/wegtam/sbt/compile/analyzer/adt/ScalacPhaseName.scala
/*
 * Copyright (C) 2017  André Schütz
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.wegtam.sbt.compile.analyzer.adt

import java.util.Locale

/**
  * Sealed trait to collect the names of the Scala compiling phases.
  */
sealed trait ScalacPhaseName

object ScalacPhaseName {
  // The phases of the Scala compiling process
  case object Loader              extends ScalacPhaseName // not official
  case object ParserPhase         extends ScalacPhaseName
  case object NamerPhase          extends ScalacPhaseName
  case object PackageObjectsPhase extends ScalacPhaseName
  case object TyperPhase          extends ScalacPhaseName
  case object PatmatPhase         extends ScalacPhaseName
  case object SuperAccessorsPhase extends ScalacPhaseName
  case object ExtMethodsPhase     extends ScalacPhaseName
  case object PicklerPhase        extends ScalacPhaseName
  case object RefChecksPhase      extends ScalacPhaseName
  case object UncurryPhase        extends ScalacPhaseName
  case object TailCallsPhase      extends ScalacPhaseName
  case object SpecializePhase     extends ScalacPhaseName
  case object ExplicitOuterPhase  extends ScalacPhaseName
  case object ErasurePhase        extends ScalacPhaseName
  case object PostErasurePhase    extends ScalacPhaseName
  case object LazyValsPhase       extends ScalacPhaseName
  case object LambdaLiftPhase     extends ScalacPhaseName
  case object ConstructorsPhase   extends ScalacPhaseName
  case object FlattenPhase        extends ScalacPhaseName
  case object MixinPhase          extends ScalacPhaseName
  case object CleanupPhase        extends ScalacPhaseName
  case object DelambdafyPhase     extends ScalacPhaseName
  case object IcodePhase          extends ScalacPhaseName
  case object JvmPhase            extends ScalacPhaseName
  case object TerminalPhase       extends ScalacPhaseName
  case object Total               extends ScalacPhaseName // not official

  /**
    * Convert the string representation of the compiler phase to the equal
    * internal object.
    *
    * @param p  String representation of the Scala compiler phase.
    * @return An option of the internal object representation.
    */
  def valueOf(p: String): Option[ScalacPhaseName] = p.toLowerCase(Locale.ROOT).trim match {
    case "loader"         => Option(Loader)
    case "parser"         => Option(ParserPhase)
    case "namer"          => Option(NamerPhase)
    case "packageobjects" => Option(PackageObjectsPhase)
    case "typer"          => Option(TyperPhase)
    case "patmat"         => Option(PatmatPhase)
    case "superaccessors" => Option(SuperAccessorsPhase)
    case "extmethods"     => Option(ExtMethodsPhase)
    case "pickler"        => Option(PicklerPhase)
    case "refchecks"      => Option(RefChecksPhase)
    case "uncurry"        => Option(UncurryPhase)
    case "tailcalls"      => Option(TailCallsPhase)
    case "specialize"     => Option(SpecializePhase)
    case "explicitouter"  => Option(ExplicitOuterPhase)
    case "erasure"        => Option(ErasurePhase)
    case "posterasure"    => Option(PostErasurePhase)
    case "lazyvals"       => Option(LazyValsPhase)
    case "lambdalift"     => Option(LambdaLiftPhase)
    case "constructors"   => Option(ConstructorsPhase)
    case "flatten"        => Option(FlattenPhase)
    case "mixin"          => Option(MixinPhase)
    case "cleanup"        => Option(CleanupPhase)
    case "delambdafy"     => Option(DelambdafyPhase)
    case "icode"          => Option(IcodePhase)
    case "jvm"            => Option(JvmPhase)
    case "terminal"       => Option(TerminalPhase)
    case "total"          => Option(Total)
    case _                => None
  }

  /**
    * Return the ID of the Scala compiler phase.
    *
    * @param p  The object representation of the Scala compiler phase.
    * @return An option to the ID (Int).
    */
  def phaseId(p: ScalacPhaseName): Option[Int] = p match {
    case Loader              => Option(0)
    case ParserPhase         => Option(1)
    case NamerPhase          => Option(2)
    case PackageObjectsPhase => Option(3)
    case TyperPhase          => Option(4)
    case PatmatPhase         => Option(5)
    case SuperAccessorsPhase => Option(6)
    case ExtMethodsPhase     => Option(7)
    case PicklerPhase        => Option(8)
    case RefChecksPhase      => Option(9)
    case UncurryPhase        => Option(10)
    case TailCallsPhase      => Option(11)
    case SpecializePhase     => Option(12)
    case ExplicitOuterPhase  => Option(13)
    case ErasurePhase        => Option(14)
    case PostErasurePhase    => Option(15)
    case LazyValsPhase       => Option(16)
    case LambdaLiftPhase     => Option(17)
    case ConstructorsPhase   => Option(18)
    case FlattenPhase        => Option(19)
    case MixinPhase          => Option(20)
    case CleanupPhase        => Option(21)
    case DelambdafyPhase     => Option(22)
    case IcodePhase          => Option(23)
    case JvmPhase            => Option(24)
    case TerminalPhase       => Option(25)
    case Total               => Option(26)
    case _                   => None
  }

}
